File: scripts/enum/rds.js
'use strict';

const {
  RDSClient,
  DescribeDBInstancesCommand,
  DescribeDBSnapshotsCommand,
  DescribeDBSnapshotAttributesCommand,
} = require('@aws-sdk/client-rds');

const { withRetry, paginate, createLogger } = require('../lib');
const { baseEnum } = require('../lib/base-enum');

// --- Findings generation ---

function generateInstanceFindings(instance) {
  const findings = [];

  if (instance.publicly_accessible) {
    findings.push({
      type: 'publicly_accessible',
      severity: 'high',
      detail: 'DB instance is publicly accessible',
    });
  }

  if (!instance.storage_encrypted) {
    findings.push({
      type: 'no_encryption',
      severity: 'high',
      detail: 'DB instance storage is not encrypted',
    });
  }

  if (!instance.iam_auth_enabled) {
    findings.push({
      type: 'no_iam_auth',
      severity: 'low',
      detail: 'IAM authentication is not enabled',
    });
  }

  if (!instance.deletion_protection) {
    findings.push({
      type: 'no_deletion_protection',
      severity: 'low',
      detail: 'Deletion protection is not enabled',
    });
  }

  return findings;
}

function generateSnapshotFindings(snapshot) {
  const findings = [];

  if (snapshot.public) {
    findings.push({
      type: 'public_snapshot',
      severity: 'critical',
      detail: 'DB snapshot is publicly accessible (shared with all)',
    });
  }

  if (!snapshot.encrypted) {
    findings.push({
      type: 'unencrypted_snapshot',
      severity: 'high',
      detail: 'DB snapshot is not encrypted',
    });
  }

  if (snapshot.shared_with && snapshot.shared_with.length > 0) {
    findings.push({
      type: 'shared_snapshot',
      severity: 'medium',
      detail: `Snapshot shared with ${snapshot.shared_with.length} account(s)`,
    });
  }

  return findings;
}

// --- Run (dependency-injectable) ---

async function run(opts = {}) {
  const runDir = opts.runDir;
  const region = opts.region;
  const accountId = opts.accountId;

  const client = opts.clients?.rds ?? new RDSClient({ region });

  const logger = opts.logger || createLogger(runDir, 'rds');
  logger.log('info', 'RDS_Enumeration_Start', { region });

  const findings = [];
  let status = 'complete';
  const errors = [];

  // --- DB Instances (Marker pagination) ---
  let allInstances;
  try {
    logger.log('api_call', 'DescribeDBInstances', { service: 'rds' });
    allInstances = await paginate(client, DescribeDBInstancesCommand, 'DBInstances', {
      tokenKey: 'Marker',
      responseTokenKey: 'Marker',
    });
  } catch (err) {
    logger.log('error', 'DescribeDBInstances', { error: err.message });
    await logger.flush();
    throw new Error(`DescribeDBInstances failed: ${err.message}`);
  }

  for (const db of allInstances) {
    const instanceFinding = {
      resource_type: 'rds_instance',
      resource_id: db.DBInstanceIdentifier,
      arn: db.DBInstanceArn,
      region,
      engine: db.Engine,
      engine_version: db.EngineVersion,
      publicly_accessible: db.PubliclyAccessible || false,
      storage_encrypted: db.StorageEncrypted || false,
      kms_key_id: db.KmsKeyId || null,
      iam_auth_enabled: db.IAMDatabaseAuthenticationEnabled || false,
      deletion_protection: db.DeletionProtection || false,
      security_groups: (db.VpcSecurityGroups || []).map((sg) => ({
        id: sg.VpcSecurityGroupId,
        status: sg.Status,
      })),
      multi_az: db.MultiAZ || false,
      findings: [],
    };

    instanceFinding.findings = generateInstanceFindings(instanceFinding);
    findings.push(instanceFinding);
  }

  // --- DB Snapshots (manual only, Marker pagination) ---
  let allSnapshots;
  try {
    logger.log('api_call', 'DescribeDBSnapshots', { service: 'rds', type: 'manual' });
    allSnapshots = await paginate(client, DescribeDBSnapshotsCommand, 'DBSnapshots', {
      params: { SnapshotType: 'manual' },
      tokenKey: 'Marker',
      responseTokenKey: 'Marker',
    });
  } catch (err) {
    errors.push({ resource: 'snapshots', error: err.message });
    logger.log('error', 'DescribeDBSnapshots', { error: err.message });
    allSnapshots = [];
    status = 'partial';
  }

  for (const snap of allSnapshots) {
    const snapshotFinding = {
      resource_type: 'rds_snapshot',
      resource_id: snap.DBSnapshotIdentifier,
      arn: snap.DBSnapshotArn,
      region,
      engine: snap.Engine,
      encrypted: snap.Encrypted || false,
      kms_key_id: snap.KmsKeyId || null,
      db_instance_identifier: snap.DBInstanceIdentifier,
      public: false,
      shared_with: [],
      findings: [],
    };

    // DescribeDBSnapshotAttributes — check for public/shared
    try {
      logger.log('api_call', 'DescribeDBSnapshotAttributes', { snapshot: snap.DBSnapshotIdentifier });
      const attrResp = await withRetry(() =>
        client.send(new DescribeDBSnapshotAttributesCommand({
          DBSnapshotIdentifier: snap.DBSnapshotIdentifier,
        }))
      );
      const attrResult = attrResp.DBSnapshotAttributesResult;
      if (attrResult && attrResult.DBSnapshotAttributes) {
        for (const attr of attrResult.DBSnapshotAttributes) {
          if (attr.AttributeName === 'restore') {
            const values = attr.AttributeValues || [];
            if (values.includes('all')) {
              snapshotFinding.public = true;
            }
            // Account IDs that are not 'all'
            snapshotFinding.shared_with = values.filter((v) => v !== 'all');
          }
        }
      }
    } catch (err) {
      errors.push({ resource: snap.DBSnapshotIdentifier, error: err.message });
      logger.log('warning', 'DescribeDBSnapshotAttributes', {
        snapshot: snap.DBSnapshotIdentifier,
        error: err.message,
      });
    }

    snapshotFinding.findings = generateSnapshotFindings(snapshotFinding);
    findings.push(snapshotFinding);
  }

  if (errors.length > 0 && status === 'complete') status = 'partial';

  await logger.flush();
  return { findings, status };
}

if (require.main === module) {
  baseEnum({ module: 'rds', run });
}

module.exports = { run };

File: scope/enumerators/rds.py
from __future__ import annotations

from typing import Any

from scope.core.coverage import CoverageTracker
from scope.core.models import ModuleEnvelope


PRIMARY_CHECKS = ["describe_db_instances"]
REQUIRED_CHECKS = ["describe_db_snapshots", "snapshot_attributes"]


def run(factory: Any, region: str) -> ModuleEnvelope:
    tracker = CoverageTracker(primary=PRIMARY_CHECKS, required=REQUIRED_CHECKS)
    rds = factory.client("rds")
    findings: list[dict[str, Any]] = []

    try:
        instances = factory.paginate(rds, "describe_db_instances", "DBInstances")
        tracker.record_ok("describe_db_instances", resource="db_instances")
    except Exception as err:
        tracker.record_module_failure("describe_db_instances", operation="rds.DescribeDBInstances", err=err)
        coverage, errors = tracker.to_envelope_fields()
        return ModuleEnvelope(
            module="rds",
            account_id=factory.account_id,
            region=region,
            status=tracker.derive_status(),
            resources=[],
            coverage=coverage,
            errors=errors,
        )

    for instance in instances:
        finding = _instance_finding(instance, region)
        finding["findings"] = _generate_instance_findings(finding)
        findings.append(finding)

    snapshots: list[dict[str, Any]] = []
    try:
        snapshots = factory.paginate(rds, "describe_db_snapshots", "DBSnapshots", SnapshotType="manual")
        tracker.record_ok("describe_db_snapshots", resource="manual_snapshots")
    except Exception as err:
        tracker.record_resource_failure(
            "describe_db_snapshots",
            resource=None,
            operation="rds.DescribeDBSnapshots",
            err=err,
        )

    for snapshot in snapshots:
        snapshot_arn = snapshot.get("DBSnapshotArn")
        finding = {
            "resource_type": "rds_snapshot",
            "resource_id": snapshot.get("DBSnapshotIdentifier"),
            "arn": snapshot_arn,
            "region": region,
            "engine": snapshot.get("Engine"),
            "encrypted": snapshot.get("Encrypted") or False,
            "kms_key_id": snapshot.get("KmsKeyId"),
            "db_instance_identifier": snapshot.get("DBInstanceIdentifier"),
            "public": False,
            "shared_with": [],
            "snapshot_attributes_status": None,
            "findings": [],
        }

        try:
            response = rds.describe_db_snapshot_attributes(DBSnapshotIdentifier=snapshot.get("DBSnapshotIdentifier"))
            result = response.get("DBSnapshotAttributesResult") or {}
            for attribute in result.get("DBSnapshotAttributes") or []:
                if attribute.get("AttributeName") == "restore":
                    values = attribute.get("AttributeValues") or []
                    finding["public"] = "all" in values
                    finding["shared_with"] = [value for value in values if value != "all"]
            finding["snapshot_attributes_status"] = "present"
            tracker.record_ok("snapshot_attributes", resource=snapshot_arn)
        except Exception as err:
            code = _error_code(err)
            finding["snapshot_attributes_status"] = _classify_status(code)
            tracker.record_resource_failure(
                "snapshot_attributes",
                resource=snapshot_arn,
                operation="rds.DescribeDBSnapshotAttributes",
                err=err,
            )

        finding["findings"] = _generate_snapshot_findings(finding)
        findings.append(finding)

    coverage, errors = tracker.to_envelope_fields()
    return ModuleEnvelope(
        module="rds",
        account_id=factory.account_id,
        region=region,
        status=tracker.derive_status(),
        resources=findings,
        coverage=coverage,
        errors=errors,
    )


def _instance_finding(instance: dict[str, Any], region: str) -> dict[str, Any]:
    return {
        "resource_type": "rds_instance",
        "resource_id": instance.get("DBInstanceIdentifier"),
        "arn": instance.get("DBInstanceArn"),
        "region": region,
        "engine": instance.get("Engine"),
        "engine_version": instance.get("EngineVersion"),
        "publicly_accessible": instance.get("PubliclyAccessible") or False,
        "storage_encrypted": instance.get("StorageEncrypted") or False,
        "kms_key_id": instance.get("KmsKeyId"),
        "iam_auth_enabled": instance.get("IAMDatabaseAuthenticationEnabled") or False,
        "deletion_protection": instance.get("DeletionProtection") or False,
        "security_groups": [
            {
                "id": security_group.get("VpcSecurityGroupId"),
                "status": security_group.get("Status"),
            }
            for security_group in instance.get("VpcSecurityGroups") or []
        ],
        "multi_az": instance.get("MultiAZ") or False,
        "findings": [],
    }


def _generate_instance_findings(instance: dict[str, Any]) -> list[dict[str, str]]:
    findings: list[dict[str, str]] = []
    if instance["publicly_accessible"]:
        findings.append(
            {"type": "publicly_accessible", "severity": "high", "detail": "DB instance is publicly accessible"}
        )
    if not instance["storage_encrypted"]:
        findings.append({"type": "no_encryption", "severity": "high", "detail": "DB instance storage is not encrypted"})
    if not instance["iam_auth_enabled"]:
        findings.append({"type": "no_iam_auth", "severity": "low", "detail": "IAM authentication is not enabled"})
    if not instance["deletion_protection"]:
        findings.append(
            {"type": "no_deletion_protection", "severity": "low", "detail": "Deletion protection is not enabled"}
        )
    return findings


def _generate_snapshot_findings(snapshot: dict[str, Any]) -> list[dict[str, str]]:
    findings: list[dict[str, str]] = []
    if snapshot["public"]:
        findings.append(
            {
                "type": "public_snapshot",
                "severity": "critical",
                "detail": "DB snapshot is publicly accessible (shared with all)",
            }
        )
    if not snapshot["encrypted"]:
        findings.append({"type": "unencrypted_snapshot", "severity": "high", "detail": "DB snapshot is not encrypted"})
    if snapshot["shared_with"]:
        findings.append(
            {
                "type": "shared_snapshot",
                "severity": "medium",
                "detail": f"Snapshot shared with {len(snapshot['shared_with'])} account(s)",
            }
        )
    return findings


def _classify_status(code: str) -> str:
    return "access_denied" if code in {"AccessDenied", "AccessDeniedException"} else "error"


def _error_code(err: Exception) -> str:
    response = getattr(err, "response", None)
    if isinstance(response, dict):
        error = response.get("Error")
        if isinstance(error, dict) and error.get("Code"):
            return str(error["Code"])
    code = getattr(err, "Code", None)
    if code:
        return str(code)
    return err.__class__.__name__ or "Error"
